Document MSBuild Logging (#10158)
dotnet/msbuild: documentation/wiki/Logging-Internals.md

# Logging Internals

MSBuild allows to plug inbox and custom loggers - those can be registered via the API or CLI and will be receiving rich tracing information during the build process (here as well referred as `BuildEventArgs`). Detailed article from user point of view can be found on Microsoft Learn: [Build Loggers](https://learn.microsoft.com/en-us/visualstudio/msbuild/build-loggers)

> Logging architecture allows usage in distributed execution nature of MSBuild (as [MSBuild has multi-process execution modes](Nodes-Orchestration.md#need-for-multiple-processes)) while pluggable logger might decide to be aware of the situation (then we speak about so called '*Distributed Loggers*') or they can operate in a mode where that fact is completely transparent to them (they can be referred to as '*Central Loggers*'). Detailed article from user point of view can be found on Microsoft Learn: [Write multi-processor-aware loggers](https://learn.microsoft.com/en-us/visualstudio/msbuild/writing-multi-processor-aware-loggers) and [Create forwarding loggers](https://learn.microsoft.com/en-us/visualstudio/msbuild/creating-forwarding-loggers)

Author: reworded to

```
(...) or they can operate in a mode where the distributed execution is not visible to them
```

Not sure if that reads better?

## Terminology

* **Logger** - an implementor of `ILogger` (or `INodeLogger`) interface. Such a component then can be registered to receive detailed tracing info from the build process and process the info further. The most common use-case is logging the information into the console, file etc. - hence the name 'Logger', but it can be processed for other purposes (e.g. the [ParallelBuildDebuggingLogger](https://github.com/rainersigwald/parallelbuilddebugginglogger) for debugging purposes, or the BuildCheck for diagnostics).
* **Central Logger** - a logger that runs in the Central node - as well here referred to as the 'Entrypoint Node'. It receives the information from the whole build regardless of if it's executed within single or multiple processes.
* **Forwarding Logger** - a logger that has a single instance in each logical executing node and is capable of filtering and/or alter the tracing stream formed of the `BuildEventArg`s before it undergoes serialization and remoting into the Entrypoint node. Main built-in implementations are: `CentralForwardingLogger` and `ConfigurableForwardingLogger`.
* **Distributed Logger** - It is a pair of Central Logger and a 'Forwarding Logger'

> * <a name="LoggerDescription"></a>**LoggerDescription** - Serializable information describing the Forwarding Logger - so that the forwarding logger can be constructed in the OOP node. It also holds the 'LoggerId' (AKA 'SinkId') - so that the remote data can be properly routed to the Central Logger. Other notable datapoints are `Verbosity` and `LoggerSwitchParameters` - which both serves for proper initialization of the forwarder after it's constructed in OOP node.

Author: Reordered

>   ![Distributed Logger](distnode.png)

Author: It is a deficiency of the PR viewing. For the actual appearance (and working links) you'd need to view it in the source branch: https://github.com/JanKrivanek/msbuild/blob/doc/logging/documentation/wiki/Logging-Internals.md

* **Sink** - Consumer of the tracing data that exposes and fires them as events - serving as a 'EventSource' for the consuming Logger
* **In-Proc node** - the build executing unit that is running in the 'Entrypoint node'
* **Out-Of-Proc node** / **OOP node** - the build executing unit that is running in the separate process.

## LoggingService

`LoggingService` is the backbone of the Logging Infrastructure - it serves as the consumer of the logging from the build process and as a dispatcher to the individual distributed and central loggers (and internally as well as x-process transmission channel for the tracing in OOP nodes).

There is a single instance in the Entrypoint node, single instance within each out-of-proc node and one in ProjectCollection (for standalone API driven evaluation). So, we can technically treat `LoggingService` as a singleton.

### Methods overview

<a name="RegisterLogger"></a>**`RegisterLogger(ILogger logger)`**

* Registers logger as distributed with a single CentralForwardingLogger (regardless of number of calls to the method - just single forwarder)
* Called by BuildManager when starting the build (on the entrypoint node)

> * Called by ProjectCollection to register loggers for operations through OM/API. ProjectCollection registers the loggers wrapped by `ReusableLogger` (which demultiplexes separate design time and build time invocation build events into a single registered logger)

Author: Good point - replaced with 'combines'

 #### OOP Node 

The initial part of the workflow of logging is identical to the in-proc delivery:

* There is 1 instance of LoggingService per build.
* It holds 1 instance of EventSourceSink pumping events to forwarders.
* There can be multiple forwarders (one shared for all central loggers registered via [`RegisterLogger`](#RegisterLogger) and one per each distributed logger registered via [`RegisterDistributedLogger`](#RegisterDistributedLogger)).
Example of forwarders:
  * `CentralForwardingLogger`
  * `ConfigurableForwardingLogger`
* Each forwarder processes the trace stream and forwards to an `EventRedirectorToSink` (1 redirector per 1 forwarder)

Processing from `EventRedirectorToSink` differs. This is given by injection of a single sink `BuildEventArgTransportSink` (injected via [`InitializeNodeLoggers`](#InitializeNodeLoggers)) as a consumer of `EventRedirectorToSink` data:

> * There is single `BuildEventArgTransportSink` per build, that consumes data from all `EventRedirectorToSink`. The data are enriched with the SinkId (AKA LoggerId) - the Id was created during the Logger registration and remoted to the OOP node together with the [`LoggerDescription`](#LoggerDescription).

Reviewer: Is there a difference betwwe SinkId and LoggerID? Or is it just different names for the same thing?

Author: It is the same thing. But different components in code refer to it via those 2 different terms - so I wanted to make it explicit here that it's the same thing

* `BuildEventArgTransportSink` is passed to the LoggingService - it is attached to each forwarding logger together with the central logger id, within a `EventRedirectorToSink`
* `BuildEventArgTransportSink` takes care about bundling the build events together with the target logger id (in some contexts referred as sinkId) and sending them through node -> node communication
	
<a name="ProcessLoggingEvent"></a>**`ProcessLoggingEvent(object buildEvent)`**
* Ingestion point of build events
* Can be called either to consume an event logged through the internal API (from LogBuildEvent or Log<XXX> [e.g. `LogBuildStarted`, `LogError` etc.])
* Or is called from the node -> node remoting (via PacketReceived)
* Ensures isolated delivery - as in synchronous mode it locks, in asynchronous mode it enqueues and delivers from the queue via single thread
* The isolated delivery however means that a single slow logger can block all others and that the event consuming time is a sum of all loggers consuming times.
 
 

>  ### Delivery of events in a single node build

Reviewer 2: ```suggestion
 ### Delivery of events in a single-node build
```

* There is 1 instance of LoggingService per build.
* It holds 1 instance of EventSourceSink pumping events to forwarders.
* There can be multiple forwarders (one shared for all central loggers registered via [`RegisterLogger`](#RegisterLogger) and one per each distributed logger registered via [`RegisterDistributedLogger`](#RegisterDistributedLogger)).
Example of forwarders:
  * `CentralForwardingLogger`
  * `ConfigurableForwardingLogger`

> * Each forwarder process the trace stream and forwards to an `EventRedirectorToSink` (1 redirector per 1 forwarder)

Reviewer 2: ```suggestion
* Each forwarder process the trace stream and may choose to forward to an `EventRedirectorToSink` (1 redirector per 1 forwarder)
```

 In case of distributed execution of build, NodeManager is requesting execution of additional Out Of Process (OOP) Nodes. Logging events from those nodes are transferred to the Entrypoint node, where they are eventually consumed by the Central Loggers. In addition to the OOP Nodes, build can have a single 'In Proc Node' - an execution node sharing the same process with the Entrypoint Node. For this reason the described process and diagram for the [Delivery of events in a single node build](#delivery-of-events-in-a-single-node-build) can apply fully for the distributed build as well (in addition to the following).

 #### OOP Node 

The initial part of the workflow of logging is identical to the in-proc delivery:

* There is 1 instance of LoggingService per build.
* It holds 1 instance of EventSourceSink pumping events to forwarders.
* There can be multiple forwarders (one shared for all central loggers registered via [`RegisterLogger`](#RegisterLogger) and one per each distributed logger registered via [`RegisterDistributedLogger`](#RegisterDistributedLogger)).
Example of forwarders:
  * `CentralForwardingLogger`
  * `ConfigurableForwardingLogger`

> * Each forwarder processes the trace stream and forwards to an `EventRedirectorToSink` (1 redirector per 1 forwarder)

Reviewer 2: ```suggestion
* Each forwarder processes the trace stream and optionally forwards each event to an `EventRedirectorToSink` (1 redirector per 1 forwarder)
```

Processing from `EventRedirectorToSink` differs. This is given by injection of a single sink `BuildEventArgTransportSink` (injected via [`InitializeNodeLoggers`](#InitializeNodeLoggers)) as a consumer of `EventRedirectorToSink` data:

* There is single `BuildEventArgTransportSink` per build, that consumes data from all `EventRedirectorToSink`. The data are enriched with the SinkId (AKA LoggerId) - the Id was created during the Logger registration and remoted to the OOP node together with the [`LoggerDescription`](#LoggerDescription).
* `BuildEventArgTransportSink` bundles the tracing data (instance of `BuildEventArgs`) together with the SinkId and passes it down the pipeline that enqueues data for remoting to the Entrypoint Node.

Simplified diagram of the flow (calls within the single component are omitted):

![OOP Node Logging](OOPNodeLoggingFlow.png)

Illustrative stack of the sequence of the calls:

![OOP Node Logging - stack](OOPLoggingStack.png)

>  #### Entrypoint Node

Reviewer 2: IMO this should be 
```suggestion
 #### Scheduler Node
```

that it is generally the entrypoint isn't the interesting part.

Author: Good point - if there is a 'thin client' in the future - this would became confusing (or even wrong). I'm replacing the term in the whole doc.

 * `LoggingService` recives the remoted data - the instance of `BuildEventArgs` bundled with the SinkId.

>  * Based on the SinkId the `LoggingService` routes the received data to the registered `EventSourceSink` (one per registered forwarder).

Reviewer 2: This doesn't make sense to me. Can you elaborate? I thought we'd forward an event if any of the distributed loggers asked for that and then on the central side pass any received event to all registered loggers. Is that not true? If not what happens if two loggers forward an event, is it serialized multiple times?

Author: If multiple distributed loggers are registered and they all request forwarding of a particular event, then it is being serialized and sent multiple times:

https://github.com/dotnet/msbuild/blob/49afd897383524da3d6e342b8254dfa8486a3400/src/Build/BackEnd/Components/Logging/BuildEventArgTransportSink.cs#L135-L151

 Simplified diagram of the flow (calls within the single component are omitted):

![EntryPoint Node Logging](EntryPointNodeLoggingFlow.png)

Illustrative stack of the sequence of the calls:

![EntryPoint Node Logging - stack](RemotedLoggingStack.png)

### Synchronous and Asynchronous events processing

As was mentioned in the [`ProcessLoggingEvent`](#ProcessLoggingEvent) description - the `LoggingService` can deliver the logging events either in a synchronous or asynchronous mode.

The logging mode is dictated by the `LoggerMode` that is injected into the `LoggingService` as such:
 * In the Out Of Proc node the mode is always **asynchronous**.
 * In the Entrypoint node the mode is by default **synchronous**, unless the `MSBUILDLOGASYNC` environment variable is set to `"1"`
 * For `ProjectCollection` the mode depends on the construction argument `useAsynchronousLogging`. In signatures that do not ask this argument - it defaults to **synchronous**.

> Regardless of the mode used - sequential and isolated delivery of events is always guaranteed (single logger will not receive next event before returning from the previous, any logger will not receive an event while it's being processed by a different logger). This behavior is of course only in the current state - it can be subject to change.

Reviewer 2: > This behavior is of course only in the current state - it can be subject to change.

Could it? I don't think violating sequential delivery ever makes sense.

Author: This was refering to the '`any logger will not receive an event while it's being processed by a different logger`' part - I'll rephrase
